torch/lib/c10d/TCPStore.cpp

```diff
@@ -591,6 +591,229 @@ void TCPStoreMasterDaemon::run() {
+    LOG(WARNING)
+        << "The multi-tenant feature of TCPStore is not implemented yet.";
+  }
+
+  TCPSocket socket{};
+  PortType port{};
+
+  std::tie(socket, port) = tcputil::listen(opts.port);
+
+  auto daemon = std::make_unique<TCPStoreMasterDaemon>(socket.handle());
+
+  return std::make_shared<TCPServer>(
+      std::move(socket), port, std::move(daemon));
+}
+
+class TCPClient {
+ public:
+  static std::unique_ptr<TCPClient> connect(
+      const SocketAddress& addr,
+      const TCPStoreOptions& opts);
+
+  void sendCommand(QueryType type) {
+    tcputil::sendValue<QueryType>(socket_.handle(), type);
+  }
+
+  void sendCommandForKey(QueryType type, const std::string& key);
+
+  void sendBits(const std::vector<std::uint8_t>& value) {
+    tcputil::sendVector<std::uint8_t>(socket_.handle(), value);
+  }
+
+  void sendStrings(c10::ArrayRef<std::string> value);
+
+  template <typename T>
+  void sendValue(const T& value) {
+    tcputil::sendValue<T>(socket_.handle(), value);
+  }
+
+  std::vector<std::uint8_t> receiveBits() {
+    return tcputil::recvVector<std::uint8_t>(socket_.handle());
+  }
+
+  template <typename T>
+  T receiveValue() {
+    return tcputil::recvValue<T>(socket_.handle());
+  }
+
+  void setTimeout(std::chrono::milliseconds value);
+
+  explicit TCPClient(TCPSocket&& socket) : socket_{std::move(socket)} {}
+
+ private:
+  TCPSocket socket_;
+};
+
+std::unique_ptr<TCPClient> TCPClient::connect(
+    const SocketAddress& addr,
+    const TCPStoreOptions& opts) {
+  TCPSocket socket =
+      tcputil::connect(addr.host, addr.port, /* wait */ true, opts.timeout);
+
+  return std::make_unique<TCPClient>(std::move(socket));
+}
+
+void TCPClient::sendCommandForKey(QueryType type, const std::string& key) {
+  tcputil::sendValue<QueryType>(socket_.handle(), type);
+
+  bool withValue = type == QueryType::SET || type == QueryType::COMPARE_SET ||
+      type == QueryType::ADD;
+
+  tcputil::sendString(socket_.handle(), key, withValue);
+}
+
+void TCPClient::sendStrings(c10::ArrayRef<std::string> value) {
+  std::size_t size = value.size();
+
+  tcputil::sendBytes<std::size_t>(socket_.handle(), &size, 1, size > 0);
+
+  if (value.empty()) {
+    return;
+  }
+
+  for (auto pos = value.begin(), last = value.end() - 1; pos <= last; ++pos) {
+    tcputil::sendString(socket_.handle(), *pos, pos != last);
+  }
+}
+
+void TCPClient::setTimeout(std::chrono::milliseconds value) {
+  if (value == std::chrono::milliseconds::zero()) {
+    return;
+  }
+
+#ifdef _WIN32
+  struct timeval timeoutTV = {value.count() / 1000,
+                              (value.count() % 1000) * 1000};
+#else
+  struct timeval timeoutTV = {.tv_sec = value.count() / 1000,
+                              .tv_usec = (value.count() % 1000) * 1000};
+#endif
+  SYSCHECK_ERR_RETURN_NEG1(::setsockopt(
+      socket_.handle(),
+      SOL_SOCKET,
+      SO_RCVTIMEO,
+      reinterpret_cast<char*>(&timeoutTV),
+      sizeof(timeoutTV)));
+}
+
+class TCPCallbackClient {
```

Comment:
Is there any reason for creating a dedicated `TCPCallbackClient` class, instead making this as a feature on TCPClient?

cc @H-Huang

Reply:
I refactored the code in a way that each socket is abstracted into its own type (`TCPServer`, `TCPClient`, `TCPCallbackClient`) to make sure I maintain single responsibility principle for each type. Since the callback mechanism uses its own socket, I ended up with `TCPCallbackClient`.

Reply:
This reasoning makes sense to me